packages/network-controller/src/NetworkController.ts

```diff
@@ -896,6 +1003,234 @@ export class NetworkController extends BaseControllerV2<
+        type: NetworkClientType.Infura,
+        network,
+        infuraProjectId: this.#infuraProjectId,
+      };
+      return [network, networkClientConfiguration];
+    });
+  }
+
+  /**
+   * Constructs the list of network clients for custom networks (that is, those
+   * that have been added to `networkConfigurations` in state).
+   *
+   * @returns The network clients.
+   */
+  #buildIdentifiedCustomNetworkClientConfigurations(): [
+    string,
+    CustomNetworkClientConfiguration,
+  ][] {
+    return Object.entries(this.state.networkConfigurations).map(
+      ([id, networkConfiguration]) => {
+        const networkClientConfiguration: CustomNetworkClientConfiguration = {
+          type: NetworkClientType.Custom,
+          chainId: networkConfiguration.chainId,
+          rpcUrl: networkConfiguration.rpcUrl,
+        };
+        return [id, networkClientConfiguration];
+      },
+    );
+  }
+
+  /**
+   * Converts the provider config object in state to a network client
+   * configuration object.
+   *
+   * @returns The network client config.
+   * @throws If the provider config is of type "rpc" and lacks either a
+   * `chainId` or an `rpcUrl`.
+   */
+  #buildIdentifiedNetworkClientConfigurationFromProviderConfig(): [
+    string,
+    NetworkClientConfiguration,
+  ] {
+    const { providerConfig } = this.state;
+
+    if (providerConfig.type === NetworkType.rpc) {
+      if (providerConfig.chainId === undefined) {
+        throw new Error('chainId must be provided for custom RPC endpoints');
+      }
+      if (providerConfig.rpcUrl === undefined) {
+        throw new Error('rpcUrl must be provided for custom RPC endpoints');
+      }
+      const id =
+        providerConfig.id ??
+        [providerConfig.chainId, providerConfig.rpcUrl].join('||');
+      const networkClientConfiguration: CustomNetworkClientConfiguration = {
+        chainId: providerConfig.chainId,
+        rpcUrl: providerConfig.rpcUrl,
+        type: NetworkClientType.Custom,
+      };
+      return [id, networkClientConfiguration];
+    }
+
+    if (isInfuraProviderType(providerConfig.type)) {
+      const id = providerConfig.type;
+      const networkClientConfiguration: InfuraNetworkClientConfiguration = {
+        network: providerConfig.type,
+        infuraProjectId: this.#infuraProjectId,
+        type: NetworkClientType.Infura,
+      };
+      return [id, networkClientConfiguration];
+    }
+
+    throw new Error(`Unrecognized network type: '${providerConfig.type}'`);
+  }
+
+  /**
+   * Uses the information in the provider config object to look up a known and
+   * preinitialized network client. Once a network client is found, updates the
+   * provider and block tracker proxy to point to those from the network client,
+   * then finally creates an EthQuery that points to the provider proxy.
+   *
+   * @throws If no network client could be found matching the current provider
+   * config.
+   */
+  #applyNetworkSelection() {
```

Comment:
This method is somewhat involved. I may split this up, but I'll wait for further comments first.

Reply:
Update: This isn't horrible. I may extract an AutoManagedNetworkRegistry later if it makes sense, but I can live with this for now.

Reply:
Yeah I didn't find it too bad at all